src/codegen/buffering_consumer.cpp

```diff
@@ -0,0 +1,196 @@
+//===----------------------------------------------------------------------===//
+//
+//                         Peloton
+//
+// buffering_consumer.cpp
+//
+// Identification: src/codegen/buffering_consumer.cpp
+//
+// Copyright (c) 2015-17, Carnegie Mellon University Database Group
+//
+//===----------------------------------------------------------------------===//
+
+#include "codegen/buffering_consumer.h"
+
+#include "codegen/values_runtime_proxy.h"
+#include "codegen/value_proxy.h"
+#include "container/lock_free_array.h"
+#include "planner/binding_context.h"
+
+namespace peloton {
+namespace codegen {
+
+//===----------------------------------------------------------------------===//
+// WRAPPED TUPLE
+//===----------------------------------------------------------------------===//
+
+// Basic Constructor
+WrappedTuple::WrappedTuple() : ContainerTuple(&tuple_), tuple_() {}
+WrappedTuple::WrappedTuple(type::Value *vals, uint32_t num_vals)
+    : ContainerTuple(&tuple_), tuple_(vals, vals + num_vals) {}
+
+// Copy Constructor
+WrappedTuple::WrappedTuple(const WrappedTuple &o)
+    : ContainerTuple(&tuple_), tuple_(o.tuple_) {}
+
+WrappedTuple &WrappedTuple::operator=(const WrappedTuple &o) {
+  expression::ContainerTuple<std::vector<type::Value>>::operator=(o);
+  tuple_ = o.tuple_;
+  return *this;
+}
+
+bool WrappedTuple::operator==(const WrappedTuple &o) {
+  return this->EqualsNoSchemaCheck(o);
+}
+
+bool WrappedTuple::operator!=(const WrappedTuple &o) {
+  return !this->EqualsNoSchemaCheck(o);
+}
+
+//===----------------------------------------------------------------------===//
+// RESULT BUFFERING CONSUMER
+//===----------------------------------------------------------------------===//
+
+BufferingConsumer::BufferingConsumer(const std::vector<oid_t> &cols,
+                                     planner::BindingContext &context) {
+  for (oid_t col_id : cols) {
+    output_ais_.push_back(context.Find(col_id));
+  }
+  state.output = &tuples_;
+}
+
+// Append the array of values (i.e., a tuple) into the consumer's buffer of
+// output tuples.
+void BufferingConsumer::BufferTuple(char *state, type::Value *vals,
+                                    uint32_t num_vals) {
+  BufferingState *buffer_state = reinterpret_cast<BufferingState *>(state);
+  buffer_state->output->Append(WrappedTuple(vals, num_vals));
+}
+
+// Get a proxy to BufferingConsumer::BufferTuple(...)
+llvm::Function *BufferingConsumer::_BufferTupleProxy::GetFunction(
+    CodeGen &codegen) {
+  const std::string &fn_name =
+#ifdef __APPLE__
+      "_ZN7peloton7codegen17BufferingConsumer11BufferTupleEPcPNS_4type5ValueEj";
```

Comment:
if peloton does not support OSX, then better leave out this mingled function name instead of use the same name as Linux, or give a simple string to explicit mark unsupported platform and let the application fail.